### ISSUE TITLE
Connection forbidden error causes inconsistent pad state and data loss 

### ISSUE THREAD

**Issue by julianrother** [hedgedoc/hedgedoc#3894]
### Description

Several hours of changes to a pad were lost after reloading the page. Some of the connected users reported not seeing some of the changes (while being "online") or reported garbled pad content. HedgeDoc did not save any revisions during this period. At the time of the last revision, HedgeDoc logged the first error message (see Logs).

### Steps to reproduce

Browser A is logged into Hedgedoc, browser B is not logged in (guest access).

1. Open a pad with "freely" permission in both browsers, wait until it is fully loaded and online
2. Browser B: Use Developer Tools to enable offline mode and wait until the pad is offline
    - Note that in Chrome the Developer Tools Network tab must be open on page load so Websocket connections are properly cut off
3. Browser A: Change pad permission to "restricted"
4. Browser B: Disable offline mode and wait until HedgeDoc connects, receives the permission error and redirects to the index page
    - At this point HedgeDoc logs a "connection forbidden" error
5. Browser A: Add some text to the pad
    - Note that the live preview is not updated (in side-by-side view)
6. Browser A: Reload the page
    - Note that the changes are missing and the pad has the content from before the "connection forbidden" error

If a third (authenticated) browser is connected to the pad, the behaviour gets weirder: After the "connection forbidden" error both browsers are "online" and allow changes to the pad, but the changes are not visible in the other browser. If one reloads the page, cursor movements and changes are visible again, however, the pad content is inconsistent. Some changes are visible, others are not. Some changes cause a "Both operations have to have the same base length" error. The live preview updates on changes, but its content is inconsistent with the Markdown input.

### Expected behaviour

The permission error should not interrupt and break the other existing (and permitted) connections to the pad.

### Logs

```shell
2023-04-24T14:05:04.699Z error:         connection forbidden
2023-04-24T15:17:00.158Z error:         connection forbidden
2023-04-24T15:34:26.965Z error:         Both operations have to have the same base length
```

### Your Setup

- Host OS: Debian 11
- NodeJS version: v16.20.0
- HedgeDoc version: 1.9.7

### Additional context

The steps in "Steps to reproduce" are my attempt at reproducing a "connection forbidden" error. In our case the pad's permission was not changed at all. So I'm not sure what caused the error on our system. My best guess is a session timeout while having the pad open followed by a reconnect due to a bad internet connection.

This is the first time in serveral years that users of our system reported data loss due to this error, although we see this error message about once per month.

This issue might be related to https://github.com/hedgedoc/hedgedoc/issues/1992.

---

I looked into the code where the "connection forbidden" error originates from (in `lib/realtime.js`):

```js
function finishConnection (socket, noteId, socketId) {
  // if no valid info provided will drop the client
  if (!socket || !notes[noteId] || !users[socketId]) {
    return interruptConnection(socket, noteId, socketId)
  }
  // check view permission
  if (!checkViewPermission(socket.request, notes[noteId])) {
    interruptConnection(socket, noteId, socketId)
    return failConnection(403, 'connection forbidden', socket)
  }
  // [...]
```

I believe the `interruptConnection` function is what causes this bug:

```js
function interruptConnection (socket, noteId, socketId) {
  if (notes[noteId]) delete notes[noteId]
  if (users[socketId]) delete users[socketId]
  if (socket) { clearSocketQueue(connectionSocketQueue, socket) } else { connectionSocketQueue.shift() }
  connectNextSocket()
}
```

`interruptConnection` is called when a SocketIO connection is established without having the permission to access a pad. It cleans up the user/socket state, but it also cleans up the state for the pad in the `notes` global.

For comparison, the `disconnect` function checks if there are other connected users before deleting `notes[noteId]`:

```js
function disconnect (socket) {
    /* ... */
    // remove note in notes if no user inside
    if (Object.keys(note.users).length <= 0) {
      if (note.server.isDirty) {
        updateNote(note, function (err, _note) {
          if (err) return logger.error('disconnect note failed: ' + err)
          // clear server before delete to avoid memory leaks
          note.server.document = ''
          note.server.operations = []
          delete note.server
          delete notes[noteId]
    /* ... */
```

I guess adding such a check to `interruptConnection` would fix the issue, but I did not verify that.

As far as I understand, both updating the pad state in the database and creating revisions relies on the `notes` global. So removing the pad state from `notes` prematurely could explain the behaviour.

**Comment by ErikMichelson** [comment]
Thanks a lot for your detailed bug report and the time you invested in researching this issue! :heart:

The suspected error in the code sounds very plausible to this problem, especially as I can reproduce the situation with your description.

As you already found the relevant code parts, feel free to create a pull request fixing this problem. If you don't have the time for this or simply don't want that to do, just give a little message and we'll take a deeper look into that.

**Comment by ErikMichelson** [comment]
This has been fixed in #3924.
